### ISSUE TITLE
BUG: "TypeError: str expected, not int" when installing with pipenv on MacOS

### ISSUE THREAD

**Issue by Deuchnord** [numpy/numpy#17987]
Not sure that this issue is related to NumPy.
When I try to install NumPy with Pipenv, it fails with a `TypeError` exception.
This issue doesn't seem to happen if using `pip` directly.

### Steps to reproduce:

I could reproduce with this config:
- macOS 11.0 (Intel)
- Python 3.9.0
- pip 20.3.1
- pipenv, version 2020.11.15

```console
$ mkdir my_project && cd my_project
$ pipenv --three
$ pipenv install numpy
```

### Error message:

```
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/pipenv/patched/notpip/_vendor/pep517/_in_process.py", line 257, in <module>
    main()
  File "/usr/local/lib/python3.9/site-packages/pipenv/patched/notpip/_vendor/pep517/_in_process.py", line 240, in main
    json_out['return_val'] = hook(**hook_input['kwargs'])
  File "/usr/local/lib/python3.9/site-packages/pipenv/patched/notpip/_vendor/pep517/_in_process.py", line 110, in prepare_metadata_for_build_wheel
    return hook(metadata_directory, config_settings)
  File "/private/var/folders/6n/lncv_nl546ndn321m0fgdsfm0000gn/T/pip-build-env-uhf9fd_v/overlay/lib/python3.9/site-packages/setuptools/build_meta.py", line 157, in prepare_metadata_for_build_wheel
    self.run_setup()
  File "/private/var/folders/6n/lncv_nl546ndn321m0fgdsfm0000gn/T/pip-build-env-uhf9fd_v/overlay/lib/python3.9/site-packages/setuptools/build_meta.py", line 248, in run_setup
    super(_BuildMetaLegacyBackend,
  File "/private/var/folders/6n/lncv_nl546ndn321m0fgdsfm0000gn/T/pip-build-env-uhf9fd_v/overlay/lib/python3.9/site-packages/setuptools/build_meta.py", line 142, in run_setup
    exec(compile(code, __file__, 'exec'), locals())
  File "setup.py", line 508, in <module>
    setup_package()
  File "setup.py", line 500, in setup_package
    setup(**metadata)
  File "/private/var/folders/6n/lncv_nl546ndn321m0fgdsfm0000gn/T/pip-resolver-wlds30r_/numpy/numpy/distutils/core.py", line 169, in setup
    return old_setup(**new_attr)
  File "/private/var/folders/6n/lncv_nl546ndn321m0fgdsfm0000gn/T/pip-build-env-uhf9fd_v/overlay/lib/python3.9/site-packages/setuptools/__init__.py", line 165, in setup
    return distutils.core.setup(**attrs)
  File "/usr/local/Cellar/python@3.9/3.9.0_5/Frameworks/Python.framework/Versions/3.9/lib/python3.9/distutils/core.py", line 148, in setup
    dist.run_commands()
  File "/usr/local/Cellar/python@3.9/3.9.0_5/Frameworks/Python.framework/Versions/3.9/lib/python3.9/distutils/dist.py", line 966, in run_commands
    self.run_command(cmd)
  File "/usr/local/Cellar/python@3.9/3.9.0_5/Frameworks/Python.framework/Versions/3.9/lib/python3.9/distutils/dist.py", line 985, in run_command
    cmd_obj.run()
  File "/private/var/folders/6n/lncv_nl546ndn321m0fgdsfm0000gn/T/pip-build-env-uhf9fd_v/overlay/lib/python3.9/site-packages/setuptools/command/dist_info.py", line 31, in run
    egg_info.run()
  File "/private/var/folders/6n/lncv_nl546ndn321m0fgdsfm0000gn/T/pip-resolver-wlds30r_/numpy/numpy/distutils/command/egg_info.py", line 24, in run
    self.run_command("build_src")
  File "/usr/local/Cellar/python@3.9/3.9.0_5/Frameworks/Python.framework/Versions/3.9/lib/python3.9/distutils/cmd.py", line 313, in run_command
    self.distribution.run_command(command)
  File "/usr/local/Cellar/python@3.9/3.9.0_5/Frameworks/Python.framework/Versions/3.9/lib/python3.9/distutils/dist.py", line 985, in run_command
    cmd_obj.run()
  File "/private/var/folders/6n/lncv_nl546ndn321m0fgdsfm0000gn/T/pip-resolver-wlds30r_/numpy/numpy/distutils/command/build_src.py", line 144, in run
    self.build_sources()
  File "/private/var/folders/6n/lncv_nl546ndn321m0fgdsfm0000gn/T/pip-resolver-wlds30r_/numpy/numpy/distutils/command/build_src.py", line 155, in build_sources
    self.build_library_sources(*libname_info)
  File "/private/var/folders/6n/lncv_nl546ndn321m0fgdsfm0000gn/T/pip-resolver-wlds30r_/numpy/numpy/distutils/command/build_src.py", line 288, in build_library_sources
    sources = self.generate_sources(sources, (lib_name, build_info))
  File "/private/var/folders/6n/lncv_nl546ndn321m0fgdsfm0000gn/T/pip-resolver-wlds30r_/numpy/numpy/distutils/command/build_src.py", line 378, in generate_sources
    source = func(extension, build_dir)
  File "numpy/core/setup.py", line 658, in get_mathlib_info
    st = config_cmd.try_link('int main(void) { return 0;}')
  File "/usr/local/Cellar/python@3.9/3.9.0_5/Frameworks/Python.framework/Versions/3.9/lib/python3.9/distutils/command/config.py", line 241, in try_link
    self._check_compiler()
  File "/private/var/folders/6n/lncv_nl546ndn321m0fgdsfm0000gn/T/pip-resolver-wlds30r_/numpy/numpy/distutils/command/config.py", line 80, in _check_compiler
    self.fcompiler = new_fcompiler(compiler=self.fcompiler,
  File "/private/var/folders/6n/lncv_nl546ndn321m0fgdsfm0000gn/T/pip-resolver-wlds30r_/numpy/numpy/distutils/fcompiler/__init__.py", line 880, in new_fcompiler
    compiler = get_default_fcompiler(plat, requiref90=requiref90,
  File "/private/var/folders/6n/lncv_nl546ndn321m0fgdsfm0000gn/T/pip-resolver-wlds30r_/numpy/numpy/distutils/fcompiler/__init__.py", line 851, in get_default_fcompiler
    compiler_type =  _find_existing_fcompiler(matching_compiler_types,
  File "/private/var/folders/6n/lncv_nl546ndn321m0fgdsfm0000gn/T/pip-resolver-wlds30r_/numpy/numpy/distutils/fcompiler/__init__.py", line 802, in _find_existing_fcompiler
    c.customize(dist)
  File "/private/var/folders/6n/lncv_nl546ndn321m0fgdsfm0000gn/T/pip-resolver-wlds30r_/numpy/numpy/distutils/fcompiler/__init__.py", line 526, in customize
    linker_so_flags = self.flag_vars.linker_so
  File "/private/var/folders/6n/lncv_nl546ndn321m0fgdsfm0000gn/T/pip-resolver-wlds30r_/numpy/numpy/distutils/fcompiler/environment.py", line 37, in __getattr__
    return self._get_var(name, conf_desc)
  File "/private/var/folders/6n/lncv_nl546ndn321m0fgdsfm0000gn/T/pip-resolver-wlds30r_/numpy/numpy/distutils/fcompiler/environment.py", line 53, in _get_var
    var = self._hook_handler(name, hook)
  File "/private/var/folders/6n/lncv_nl546ndn321m0fgdsfm0000gn/T/pip-resolver-wlds30r_/numpy/numpy/distutils/fcompiler/__init__.py", line 705, in _environment_hook
    return hook()
  File "/private/var/folders/6n/lncv_nl546ndn321m0fgdsfm0000gn/T/pip-resolver-wlds30r_/numpy/numpy/distutils/fcompiler/gnu.py", line 346, in get_flags_linker_so
    flags = GnuFCompiler.get_flags_linker_so(self)
  File "/private/var/folders/6n/lncv_nl546ndn321m0fgdsfm0000gn/T/pip-resolver-wlds30r_/numpy/numpy/distutils/fcompiler/gnu.py", line 136, in get_flags_linker_so
    os.environ['MACOSX_DEPLOYMENT_TARGET'] = target
  File "/usr/local/Cellar/python@3.9/3.9.0_5/Frameworks/Python.framework/Versions/3.9/lib/python3.9/os.py", line 684, in __setitem__
    value = self.encodevalue(value)
  File "/usr/local/Cellar/python@3.9/3.9.0_5/Frameworks/Python.framework/Versions/3.9/lib/python3.9/os.py", line 756, in encode
    raise TypeError("str expected, not %s" % type(value).__name__)
TypeError: str expected, not int
```

**Comment by BvB93** [comment]
This same issue was previously noted in https://github.com/numpy/numpy/pull/17906.
As aforementioned PR attempts to fix the issue I'd recommend keeping an eye on it.

**Comment by Deuchnord** [comment]
Looks like the issue has been fixed in 1ccb4c6dbfa6194d1627885e39f81823bce44fc7, hasn't it?

**Comment by seberg** [comment]
Thanks for the ping @Deuchnord, yes it was.

**Comment by Deuchnord** [comment]
Thanks for your answer @seberg!
Is there any chance to release a patch version to fix the issue? It's really blocking me on my project for the Python 3.9 compatibility 😞

**Comment by mattip** [comment]
You can use the 1.20.0rc2 release candidate.
`python -m pip install -i https://pypi.anaconda.org/multibuild-wheels-staging/simple numpy==1.20.0rc2`

This should become the 1.20.0 release in a few weeks.

**Comment by rgommers** [comment]
This should be picking up a `1.19.x` wheel though - maybe `pip` too old?

**Comment by bernardopacini** [comment]
Hello,

I am running into an issue downstream of this typeerror. I am using f2py and f90wrap to python wrap a fortran code, but I am no longer able to run f2py successfully. I encounter the following error, which comes from numpy:

```
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/private/var/folders/82/x2lgxrz96clgm5p30pp_jxl40000gn/T/pip-req-build-h0n_j19k/setup.py", line 55, in <module>
        setup(name='f90wrap',
      File "/usr/local/lib/python3.9/site-packages/numpy/distutils/core.py", line 169, in setup
        return old_setup(**new_attr)
      File "/usr/local/lib/python3.9/site-packages/setuptools/__init__.py", line 153, in setup
        return distutils.core.setup(**attrs)
      File "/usr/local/Cellar/python@3.9/3.9.1_6/Frameworks/Python.framework/Versions/3.9/lib/python3.9/distutils/core.py", line 148, in setup
        dist.run_commands()
      File "/usr/local/Cellar/python@3.9/3.9.1_6/Frameworks/Python.framework/Versions/3.9/lib/python3.9/distutils/dist.py", line 966, in run_commands
        self.run_command(cmd)
      File "/usr/local/Cellar/python@3.9/3.9.1_6/Frameworks/Python.framework/Versions/3.9/lib/python3.9/distutils/dist.py", line 985, in run_command
        cmd_obj.run()
      File "/usr/local/lib/python3.9/site-packages/numpy/distutils/command/install.py", line 60, in run
        r = self.setuptools_run()
      File "/usr/local/lib/python3.9/site-packages/numpy/distutils/command/install.py", line 34, in setuptools_run
        return distutils_install.run(self)
      File "/usr/local/Cellar/python@3.9/3.9.1_6/Frameworks/Python.framework/Versions/3.9/lib/python3.9/distutils/command/install.py", line 546, in run
        self.run_command('build')
      File "/usr/local/Cellar/python@3.9/3.9.1_6/Frameworks/Python.framework/Versions/3.9/lib/python3.9/distutils/cmd.py", line 313, in run_command
        self.distribution.run_command(command)
      File "/usr/local/Cellar/python@3.9/3.9.1_6/Frameworks/Python.framework/Versions/3.9/lib/python3.9/distutils/dist.py", line 985, in run_command
        cmd_obj.run()
      File "/usr/local/lib/python3.9/site-packages/numpy/distutils/command/build.py", line 61, in run
        old_build.run(self)
      File "/usr/local/Cellar/python@3.9/3.9.1_6/Frameworks/Python.framework/Versions/3.9/lib/python3.9/distutils/command/build.py", line 135, in run
        self.run_command(cmd_name)
      File "/usr/local/Cellar/python@3.9/3.9.1_6/Frameworks/Python.framework/Versions/3.9/lib/python3.9/distutils/cmd.py", line 313, in run_command
        self.distribution.run_command(command)
      File "/usr/local/Cellar/python@3.9/3.9.1_6/Frameworks/Python.framework/Versions/3.9/lib/python3.9/distutils/dist.py", line 985, in run_command
        cmd_obj.run()
      File "/usr/local/lib/python3.9/site-packages/numpy/distutils/command/build_ext.py", line 269, in run
        fcompiler.customize(self.distribution)
      File "/usr/local/lib/python3.9/site-packages/numpy/distutils/fcompiler/__init__.py", line 505, in customize
        get_flags('opt', oflags)
      File "/usr/local/lib/python3.9/site-packages/numpy/distutils/fcompiler/__init__.py", line 496, in get_flags
        flags.extend(getattr(self.flag_vars, tag))
      File "/usr/local/lib/python3.9/site-packages/numpy/distutils/fcompiler/environment.py", line 40, in __getattr__
        return self._get_var(name, conf_desc)
      File "/usr/local/lib/python3.9/site-packages/numpy/distutils/fcompiler/environment.py", line 56, in _get_var
        var = self._hook_handler(name, hook)
      File "/usr/local/lib/python3.9/site-packages/numpy/distutils/fcompiler/__init__.py", line 703, in _environment_hook
        return hook()
      File "/usr/local/lib/python3.9/site-packages/numpy/distutils/fcompiler/gnu.py", line 222, in get_flags_opt
        v = self.get_version()
      File "/usr/local/lib/python3.9/site-packages/numpy/distutils/fcompiler/__init__.py", line 424, in get_version
        version = CCompiler.get_version(self, force=force, ok_status=ok_status)
      File "/usr/local/lib/python3.9/site-packages/numpy/distutils/ccompiler.py", line 90, in <lambda>
        m = lambda self, *args, **kw: func(self, *args, **kw)
      File "/usr/local/lib/python3.9/site-packages/numpy/distutils/ccompiler.py", line 649, in CCompiler_get_version
        version = matcher(output)
      File "/usr/local/lib/python3.9/site-packages/numpy/distutils/fcompiler/gnu.py", line 271, in version_match
        v = self.gnu_version_match(version_string)
      File "/usr/local/lib/python3.9/site-packages/numpy/distutils/fcompiler/gnu.py", line 73, in gnu_version_match
        raise ValueError(err + version_string)
    ValueError: A valid Fortran version was not found in this string:
    gfortran-10: warning: couldn't understand version 11
    10.2.0
    ```

My gcc / gfortran are up to date with the latest Brew versions and I installed numpy 1.20.0rc2.

Thank you in advance for any advice on how to proceed.

**Comment by Debbiepimpo** [comment]
I have the same error with this: `TypeError("str expected, not %s" % type(value).__name__)
TypeError: str expected, not tuple` IMPORTANT! I'm using Windows so, mac solutions are not helping for me...

**Comment by rgommers** [comment]
The earlier reported issues have been fixed in numpy `1.21.0`. If you still have an issue with that version installed @Debbiepimpo, can you please open a new issue with enough detail that we can investigate/reproduce?
